src/shacl/tokens.ts

```diff
@@ -0,0 +1,347 @@
+// we create only the category token and the full (un-prefixed) IRI token; the
+// prefixed token is created later (via `getShaclTokenTypes`) once we actually
+// know what the `xsd` prefix is.
+const xsdUnprefixedTokenMap = xsdLocalNames.reduce((tokenMap, localName) => {
+  const tokenName = `SHACL_xsd_${localName}`; // category token name
+  const iriTokenName = `${tokenName}_IRI`; // IRI token name
+  // Category token that will ultimately select either an XSD IRI or an XSD PN_LOCAL:
+  const iriOrPrefixCategoryToken = createToken({
+    name: tokenName,
+    pattern: Lexer.NA,
+  });
+
+  return {
+    ...tokenMap,
+    [tokenName]: iriOrPrefixCategoryToken,
+    [iriTokenName]: createToken({
+      name: iriTokenName,
+      pattern: `<${xsdIriNamespace}${localName}`,
+      categories: [iriOrPrefixCategoryToken, turtleTokenMap.IRIREF],
+    }),
+  };
+}, {});
+
+// We can also pre-compute all SHACL tokens except for those that include
+// prefixes (again, since we don't know a priori what the SHACL prefix will be).
+// This helps keep our parser quick. We do it in the same way that we did for
+// XSD tokens, above -- we create a "category" token for each SHACL local name
+// that will be used to match either the full (un-prefixed) IRI or the prefixed
+// name (once we know what the SHACL prefix is).
+const shaclUnprefixedTokenMap = localNames.reduce((tokenMap, localName) => {
+  const category = localNameToCategoryMap[localName];
+  const categoryToken = categoryTokenMap[category];
+  const tokenName = `SHACL_${localName}`;
+  const iriTokenName = `${tokenName}_IRI`;
+  // Category token that will select either a SHACL IRI or a SHACL PN_LOCAL:
+  const iriOrPrefixCategoryToken = createToken({
+    name: tokenName,
+    pattern: Lexer.NA,
+    categories: categoryToken ? [categoryToken] : [],
+  });
+
+  return {
+    ...tokenMap,
+    [tokenName]: iriOrPrefixCategoryToken,
+    [iriTokenName]: createToken({
+      name: iriTokenName,
+      pattern: `<${shaclIriNamespace}${localName}>`,
+      categories: [iriOrPrefixCategoryToken, turtleTokenMap.IRIREF],
+    }),
+  };
+}, xsdUnprefixedTokenMap);
+
+const makePrefixer = (prefix: string) => (localName: string) =>
+  `${prefix}:${localName}`;
+
+type LocalName = {
+  [K in keyof typeof localNamesByCategory]: typeof localNamesByCategory[K] extends (infer T)[]
+    ? T
+    : never
+}[keyof typeof localNamesByCategory];
+
+type TokenMap = { [K in LocalName]: TokenType };
+
+// Retrieves the complete map of all SHACL/XSD tokens, given the SHACL and XSD
+// prefixes. The map contains, for every local name, a token matching the full
+// IRI, a token matching the prefixed local name, and a "category" token that
+// matches both. The category token is useful for simplifying parser rules (not
+// having to match every SHACL token as both a full IRI and a prefixed local
+// name.)
+// This function is called by the SHACL parser. It is memoized because the
```

Comment:
I'm just trying to keep up here, but where else is this called? I only see it in ShaclParser's constructor.

Reply:
I think that's the only place it's called.